gobblin-core/src/main/java/gobblin/source/extractor/extract/kafka/KafkaWrapper.java

```diff
@@ -0,0 +1,342 @@
+  }
+
+  private KafkaAPI getKafkaAPI() {
+    switch (this.apiType) {
+      case NEW:
+        return new KafkaNewAPI();
+      case OLD:
+        return new KafkaOldAPI();
+      default:
+        throw new RuntimeException(this.apiType + " is not a valid API type.");
+    }
+  }
+
+  @Override
+  public void close() throws IOException {
+    this.kafkaAPI.close();
+  }
+
+  private abstract class KafkaAPI implements Closeable {
+    protected abstract List<KafkaTopic> getFilteredTopics(Set<String> blacklist, Set<String> whitelist);
+
+    protected abstract long getEarliestOffset(KafkaPartition partition) throws KafkaOffsetRetrievalFailureException;
+
+    protected abstract long getLatestOffset(KafkaPartition partition) throws KafkaOffsetRetrievalFailureException;
+  }
+
+  /**
+   * Wrapper for the old low-level Scala-based Kafka API.
+   */
+  private class KafkaOldAPI extends KafkaAPI {
+    private static final int DEFAULT_KAFKA_TIMEOUT_VALUE = 30000;
+    private static final int DEFAULT_KAFKA_BUFFER_SIZE = 1024 * 1024;
+    private static final String DEFAULT_KAFKA_CLIENT_NAME = "kafka-old-api";
+    private static final int NUM_TRIES_FETCH_TOPIC = 3;
+    private static final int NUM_TRIES_FETCH_OFFSET = 3;
+
+    private final Map<String, SimpleConsumer> activeConsumers = Maps.newHashMap();
+
+    @Override
+    public List<KafkaTopic> getFilteredTopics(Set<String> blacklist, Set<String> whitelist) {
+      List<TopicMetadata> topicMetadataList = getFilteredMetadataList(blacklist, whitelist);
+
+      List<KafkaTopic> filteredTopics = Lists.newArrayList();
+      for (TopicMetadata topicMetadata : topicMetadataList) {
+        List<KafkaPartition> partitions = getPartitionsForTopic(topicMetadata);
+        filteredTopics.add(new KafkaTopic(topicMetadata.topic(), partitions));
+      }
+      return filteredTopics;
+    }
+
+    private List<KafkaPartition> getPartitionsForTopic(TopicMetadata topicMetadata) {
+      List<KafkaPartition> partitions = Lists.newArrayList();
+
+      for (PartitionMetadata partitionMetadata : topicMetadata.partitionsMetadata()) {
+        partitions.add(new KafkaPartition.Builder().withId(partitionMetadata.partitionId())
+            .withTopicName(topicMetadata.topic()).withLeaderId(partitionMetadata.leader().id())
+            .withLeaderHost(partitionMetadata.leader().host()).withLeaderPort(partitionMetadata.leader().port())
+            .build());
+      }
+      return partitions;
+    }
+
+    private List<TopicMetadata> getFilteredMetadataList(Set<String> blacklist, Set<String> whitelist) {
+      List<TopicMetadata> filteredTopicMetadataList = Lists.newArrayList();
+
+      //Try all brokers one by one, until successfully retrieved topic metadata (topicMetadataList is non-null)
+      for (String broker : KafkaWrapper.this.brokers) {
+        filteredTopicMetadataList = fetchTopicMetadataFromBroker(broker, blacklist, whitelist);
+        if (filteredTopicMetadataList != null) {
+          break;
```

Comment:
Instead of break, you can simply return here. Then at the end, you just need to throw the RuntimeException.